crypto/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+    fn update(&mut self, bytes: &[u8]) -> Option<()>;
+
+    /// Finish a running hash operation and return the result.
+    ///
+    /// Once this function has been called, the object can no longer be used and
+    /// a new one must be created to hash more data.
+    ///
+    /// # Arguments
+    ///
+    /// * `digest` - Where the computed digest should be written.
+    fn finish(self, digest: &mut [u8]) -> Option<()>;
+}
+
+pub trait Crypto {
+    type Cdi;
+    type Hasher: Hasher;
+
+    /// Fills the buffer with random values.
+    ///
+    /// # Arguments
+    ///
+    /// * `dst` - The buffer to be filled.
+    fn rand_bytes(dst: &mut [u8]) -> Option<()>;
+
+    /// Cryptographically hashes the given buffer.
+    ///
+    /// # Arguments
+    ///
+    /// * `profile` - Which profile is being used. This will tell the platform which algorithm to
+    ///   use.
+    /// * `bytes` - Value to be hashed.
+    /// * `digest` - Where the computed digest should be written.
+    fn hash(algs: AlgLen, bytes: &[u8], digest: &mut [u8]) -> Option<()> {
+        let mut hasher = Self::hash_initialize(algs)?;
+        hasher.update(bytes)?;
+        hasher.finish(digest)
+    }
+
+    /// Initialize a running hash. Returns an object that will be able to complete the rest.
+    ///
+    /// Used for hashing multiple buffers that may not be in consecutive memory.
+    ///
+    /// # Arguments
+    ///
+    /// * `profile` - Which profile is being used. This will tell the platform which algorithm to
+    ///   use.
+    fn hash_initialize(algs: AlgLen) -> Option<Self::Hasher>;
+
+    /// Derive a CDI based on the current base CDI and measurements.
+    ///
+    /// # Arguments
+    ///
+    /// * `profile` - Which profile is being used. This will tell the platform
+    ///   which algorithm to use
+    /// * `measurement_digest` - A digest of the measurements which should be
+    ///   used for CDI derivation
+    /// * `info` - Caller-supplied info string to use in CDI derivation
+    fn derive_cdi(algs: AlgLen, measurement_digest: &[u8], info: &[u8]) -> Option<Self::Cdi>;
+
+    /// Derives an ECDSA keypair from `cdi` and returns the public key
+    ///
+    /// # Arguments
+    ///
+    /// * `profile` - Which profile is being used. This will tell the platform
+    ///    which algorithm to use
+    /// * `cdi` - CDI from which to derive the signing key
+    /// * `label` - Caller-supplied label to use in asymmetric key derivation
+    /// * `info` - Caller-supplied info string to use in asymmetric key derivation
+    ///
+    /// Returns a derived public key
+    fn derive_ecdsa_pub(
+        algs: AlgLen,
+        cdi: &Self::Cdi,
+        label: &[u8],
+        info: &[u8],
+        pub_x: &mut [u8],
+        pub_y: &mut [u8],
+    ) -> Option<()>;
+
+    /// Sign `digest` with the platform Alias Key
+    fn ecdsa_sign_with_alias(
+        algs: AlgLen,
+        digest: &[u8],
+        sig_r: &mut [u8],
+        sig_s: &mut [u8],
```

Comment:
As a heads up, I also changed these to have mutable slice outputs because I they were relying on `DPE_PROFILE` for sizes.